homeassistant/components/prometheus/__init__.py

```diff
@@ -155,23 +159,28 @@ def __init__(
         self._metrics = {}
         self._climate_units = climate_units
 
-    def handle_state_changed(self, event):
-        """Listen for new messages on the bus, and add them to Prometheus."""
+    def handle_state_changed_event(self, event):
+        """Handle new messages from the bus."""
         if (state := event.data.get("new_state")) is None:
             return
 
-        entity_id = state.entity_id
-        _LOGGER.debug("Handling state update for %s", entity_id)
-        domain, _ = hacore.split_entity_id(entity_id)
-
         if not self._filter(state.entity_id):
+            _LOGGER.debug("Filtered out entity %s", state.entity_id)
             return
 
+        self.handle_state(state)
```

Comment:
Is there any particular reason you've moved this above the block that cleans up old metrics when an entity gets renamed? I suspect this might result in duplicate metrics being exported if Prometheus collects metrics after we create metrics for the new entity but before we clean up the old one.

Reply:
Yes, that was the reason :) There's a race here either way -- you can either have some time with missing metrics, or some time with duplicate metrics. Things really have to line up for either to trigger, but my guess was that it'd be better to have duplicate -- ICBW, happy to change it back around.

Reply:
I've changed it back in the latest update.

Reply:
Thank you! I agree there's a race either way, but I suspect a missing data point will be less surprising than a duplicate.